circuit/src/ecc/generic/ecdsa.rs

```diff
@@ -0,0 +1,219 @@
+	/// Note: it does not make sense to do this in wrong field arithmetic
+	/// because the signature requires fresh randomness (k) for security reasons so it cannot be
+	/// done in a ZK circuit.
+	pub fn sign(
+		&self, msg_hash: BigUint,
+	) -> (
+		Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+		Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+	) {
+		let mut rng = thread_rng();
+		// Draw randomness
+		let k = Fq::random(rng);
+		let k_inv = k.invert().unwrap();
+
+		// Calculate `r`
+		let r_point = (Secp256k1::generator() * k).to_affine().coordinates().unwrap();
+		let x = r_point.x();
+		let r = fp_to_fq(*x);
+		let msg_hash_fq = biguint_to_fq(msg_hash);
+		let private_key_fq = biguint_to_fq(self.private_key.value());
+		// Calculate `s`
+		let s = k_inv * (msg_hash_fq + (r * private_key_fq));
+
+		(Integer::from_w(r), Integer::from_w(s))
+	}
+}
+
+/// Struct for ECDSA verification using wrong field arithmetic.
+pub struct EcdsaVerify<N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P, EC>
+where
+	P: RnsParams<<Secp256k1Affine as CurveAffine>::Base, N, NUM_LIMBS, NUM_BITS>
+		+ RnsParams<<Secp256k1Affine as CurveAffine>::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<Secp256k1Affine>,
+{
+	phantom: std::marker::PhantomData<(N, P, EC)>,
+}
+
+impl<N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P, EC>
+	EcdsaVerify<N, NUM_LIMBS, NUM_BITS, P, EC>
+where
+	P: RnsParams<<Secp256k1Affine as CurveAffine>::Base, N, NUM_LIMBS, NUM_BITS>
+		+ RnsParams<<Secp256k1Affine as CurveAffine>::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<Secp256k1Affine>,
+{
+	/// Verify a signature for a given message hash and public key using wrong field arithmetic
+	/// Similar to the ZK circuit setting instead of computing s_inverse we take it in as an advice value
+	pub fn verify_signature_no_pubkey_check(
+		signature: (
+			Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+			Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+		),
+		msg_hash: BigUint, public_key: EcPoint<Secp256k1Affine, N, NUM_LIMBS, NUM_BITS, P, EC>,
+		advice_s_inv: Integer<Fq, N, NUM_LIMBS, NUM_BITS, P>,
+	) -> bool {
+		let r = signature.0;
+		let s = signature.1;
+		let s_mul_s_inv = s.mul(&advice_s_inv);
+		if s_mul_s_inv.result != Integer::one() {
+			return false;
+		}
+		let msg_hash_integer = Integer::<Fq, N, NUM_LIMBS, NUM_BITS, P>::new(msg_hash);
+		let u_1 = msg_hash_integer.mul(&advice_s_inv).result;
+		let u_2 = r.mul(&advice_s_inv).result;
+
+		let g = Secp256k1::generator().to_affine();
+		let g_as_ecpoint = EcPoint::<Secp256k1Affine, N, NUM_LIMBS, NUM_BITS, P, EC>::new(
+			Integer::from_w(g.x),
+			Integer::from_w(g.y),
+		);
+
+		let v_1 = g_as_ecpoint.mul_scalar(u_1);
+		let v_2 = public_key.mul_scalar(u_2);
+
+		let r_point = v_1.add(&v_2);
+		let x_candidate = r_point.x;
+		let r_mod_n = <P as RnsParams<Fq, N, NUM_LIMBS, NUM_BITS>>::compose(r.limbs);
+		let x_candidate_mod_n =
+			<P as RnsParams<Fp, N, NUM_LIMBS, NUM_BITS>>::compose(x_candidate.limbs);
```

Comment:
Is there a risk of cillision attack here?

Looks like they are checking each limb in halo2wrong: https://github.com/privacy-scaling-explorations/halo2wrong/blob/master/integer/src/chip.rs#L373-L375

Reply:
Good question...I am investigating.

Reply:
Yes, there can be a collision attack...good catch. This is because `r` is in `F_q` and `x_candidate` is in `F_p` and these two fields have different sizes. I will make the adjustments to the code.